recursion/core/src/poseidon2/external.rs

```diff
@@ -52,50 +67,80 @@ impl<F: PrimeField32> MachineAir<F> for Poseidon2Chip {
     ) -> RowMajorMatrix<F> {
-        let mut input = [F::one(); WIDTH];
-        let rows = (0..128)
-            .map(|i| {
+        let mut rows = Vec::new();
+
+        for i in 0..input.poseidon2_events.len() {
+            let event = input.poseidon2_events[i].clone();
+
+            let mut input_memory_record = event.state_read_records;
+
+            let mut pre_state: [F; WIDTH] = input_memory_record
+                .iter()
+                .take(WIDTH / 4)
+                .flat_map(|block| block.value.0)
+                .collect::<Vec<_>>()
+                .try_into()
+                .unwrap();
+
+            for i in 0..32 {
                 let mut row = [F::zero(); NUM_POSEIDON2_COLS];
                 let cols: &mut Poseidon2Cols<F> = row.as_mut_slice().borrow_mut();
-                cols.input = input;
+
+                cols.clk = event.clk + F::from_canonical_usize(i);
+
+                cols.state_ptr = event.state_ptr + F::from_canonical_usize(i * 4);
+
+                cols.pre_state
+                    .iter_mut()
+                    .zip(input_memory_record.iter())
+                    .for_each(|(pre_state_j, input_j)| {
+                        pre_state_j.populate(input_j);
+                    });
 
                 let r = i % 31;
-                let rounds_f = 8;
-                let rounds_p = 22;
-                let rounds = rounds_f + rounds_p;
-                let rounds_f_beginning = rounds_f / 2;
-                let p_end = rounds_f_beginning + rounds_p;
 
                 cols.rounds[r] = F::one();
                 let is_initial_layer = r == 0;
                 let is_external_layer = r != 0
-                    && (((r - 1) < rounds_f_beginning) || (p_end <= (r - 1) && (r - 1) < rounds));
+                    && (((r - 1) < ROUNDS_F_BEGINNING) || (P_END <= (r - 1) && (r - 1) < ROUNDS));
 
                 if is_initial_layer {
                     // Mark the selector as initial.
                     cols.is_initial = F::one();
 
-                    // Don't apply the round constants.
-                    cols.add_rc.copy_from_slice(&cols.input);
+                    // initialize the pre_state from the memory records.
+                    // Don't apply the round constants and copy the `input_pre_state` as it is.
+                    cols.add_rc.copy_from_slice(&pre_state);
                 } else if is_external_layer {
                     // Mark the selector as external.
                     cols.is_external = F::one();
 
                     // Apply the round constants.
-                    for j in 0..WIDTH {
-                        cols.add_rc[j] =
-                            cols.input[j] + F::from_wrapped_u32(RC_16_30_U32[r - 1][j]);
-                    }
+                    // Apply the round constants.
+                    cols.add_rc
+                        .iter_mut()
+                        .zip(pre_state.iter().zip(RC_16_30_U32[r - 1].iter()))
```

Comment:
why do you have to change this?

Reply:
I was making it idiomatic rust- but I will revert it to the original